### PR TITLE
Cache JIT `GroupBy.apply` functions

### PR DESCRIPTION
**Pull request by brandon-b-miller**
This PR sends incoming UDFs that go through the `engine='jit'` codepath through the main UDF cache. This should avoid recompiling if a user reuses the same UDF on different input data, so long as the types of that data are the same.

### CODE REVIEW

**Comment by bdice** [comment]
Can we merge the big move PR #12669 before doing feature work on these areas of the code?

**Comment by brandon-b-miller** [comment]
with the merge of https://github.com/rapidsai/cudf/pull/12669 this should be ready for review.

**Review by bdice — COMMENTED** [review]
Is it possible to test the caching behavior?

**Comment by brandon-b-miller** [comment]
> Is it possible to test the caching behavior?

What do you think about the approach [here](https://github.com/rapidsai/cudf/blob/branch-23.04/python/cudf/cudf/tests/test_udf_masked_ops.py#L714-L743) ?

**Comment by bdice** [comment]
> > Is it possible to test the caching behavior?
> 
> What do you think about the approach [here](https://github.com/rapidsai/cudf/blob/branch-23.04/python/cudf/cudf/tests/test_udf_masked_ops.py#L714-L743) ?

Looks good to me, that would be a good test if you copied and altered it for the groupby case.

File: python/cudf/cudf/core/udf/groupby_utils.py
Comment (bdice) [inline python/cudf/cudf/core/udf/groupby_utils.py:152]: Is there a reason we don't use `lru_cache` and instead manually track cache keys? I assume it has to do with types being supported in `lru_cache` keys?

File: python/cudf/cudf/core/udf/groupby_utils.py
Comment (bdice) [inline python/cudf/cudf/core/udf/groupby_utils.py:159]: Fill the cache if it's empty, then read the cache. This reduces duplication.
```suggestion
    if cache_key not in precompiled:
        precompiled[cache_key] = _get_groupby_apply_kernel(
            grouped_values, function, args
        )
    kernel, return_type = precompiled[cache_key]
```

**Review by bdice — COMMENTED** [review]
Test looks good but I think we can improve the cache itself.

**Review by shwina — APPROVED** [review]
Approving pending @bdice's improvement

File: python/cudf/cudf/core/udf/groupby_utils.py
Comment (brandon-b-miller) [reply python/cudf/cudf/core/udf/groupby_utils.py:152]: In this context, `precompiled` is a `cachetools.LRUCache`. Are you asking why we don't do the following from functools?
```python
@functools.lru_cache
def _get_groupby_apply_kernel(...)
```

If so the reason was that I wanted different UDF pipelines (`apply`, `groupby.apply` etc) to share the same cache.

Comment (bdice) [reply python/cudf/cudf/core/udf/groupby_utils.py:152]: Nevermind. 😄 I didn't look closely enough at `precompiled`. But to clarify, how do you distinguish the type of UDF? Could an apply function and a groupby apply function reuse the same exact kernel? If not, how are the cache keys distinguished (for functions of the same data types)?

Comment (brandon-b-miller) [reply python/cudf/cudf/core/udf/groupby_utils.py:152]: The cache key is based on the bytecode of the UDF, the particulars are found [here](https://github.com/rapidsai/cudf/blob/branch-23.04/python/cudf/cudf/utils/cudautils.py#L193). I suppose you could get a cache hit erroneously if you:
1. wrote a function `f` and executed it using `apply`
2. applied the exact same `f` on a `groupby` result whose columns were the exact same dtypes as the dataframe that it was first applied to

However I would expect the above to cause a crash in pandas case as well since each API enforces a different syntax for the kinds of UDFs it accepts, so using one kind of function with the other's `apply` API probably wouldn't work in most cases.

Comment (bdice) [reply python/cudf/cudf/core/udf/groupby_utils.py:152]: I'll defer to your judgment here -- but distinguishing keys clearly would be a plus, in my eyes. An erroneous cache hit would be bad.

**Review by bdice — APPROVED** [review]

File: python/cudf/cudf/core/udf/groupby_utils.py
Comment (brandon-b-miller) [reply python/cudf/cudf/core/udf/groupby_utils.py:152]: I added an extra tuple element to UDFs that go through `GroupBy.apply` that should break this degeneracy. 

https://github.com/rapidsai/cudf/pull/12802/commits/425a91282d8573710518208902cf7c80b23d740b

Comment (bdice) [reply python/cudf/cudf/core/udf/groupby_utils.py:152]: Nice. Thanks!

Comment (bdice) [reply python/cudf/cudf/core/udf/groupby_utils.py:152]: fwiw, I think I would have preferred an approach like `_generate_cache_key(grouped_values, function, suffix="__GROUPBY_APPLY_UDF")` where you provide a suffix to the function making the key. Not a dealbreaker but worth considering if we have more JIT code paths with separate JIT caches.

**Pull request merged**
